mavros_extras/src/plugins/mag_calibration_status.cpp

```diff
@@ -0,0 +1,102 @@
+/**
+ * @brief MagCalStatus plugin
+ * @file MagCalStatus.cpp
+ * @author André Ferreira <andre.ferreira@beyond-vision.pt>
+ *
+ * @example MagCalStatus.cpp
+ * @addtogroup plugin
+ * @{
+ */
+
+#include <mavros/mavros_plugin.h>
+#include <std_msgs/UInt8.h>
+#include <mavros_msgs/MagnetometerReporter.h>
+namespace mavros {
+namespace std_plugins {
+/**
+ * @brief MagCalStatus plugin.
+ *
+ * Example and "how to" for users.
+ */
+class MagCalStatusPlugin : public plugin::PluginBase {
+public:
+	MagCalStatusPlugin() : PluginBase(),
+		mcs_nh("~mag_calibration")
+	{ }
+
+	/**
+	 * Plugin initializer. Constructor should not do this.
+	 */
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+		mcs_pub = mcs_nh.advertise<std_msgs::UInt8>("status", 2, true);
+		mcr_pub = mcs_nh.advertise<mavros_msgs::MagnetometerReporter>("report", 2, true);
+	}
+
+	/**
+	 * This function returns message subscriptions.
+	 *
+	 * Each subscription made by PluginBase::make_handler() template.
+	 * Two variations:
+	 *  - With automatic decoding and framing error filtering (see handle_heartbeat)
+	 *  - Raw message with framig status (see handle_systemtext)
+	 */
+	Subscriptions get_subscriptions() {
+		return {
+			/* automatic message deduction by second argument */
+			make_handler(&MagCalStatusPlugin::handle_status),
+			make_handler(&MagCalStatusPlugin::handle_report),
+		};
+	}
+
+private:
+	ros::NodeHandle mcs_nh;
+	ros::Publisher mcs_pub;
+	ros::Publisher mcr_pub;
+	std::vector<bool> calibration_show;
+	std::vector<uint8_t> _rg_compass_cal_progress;
+	//Send progress of magnetometer calibration
+	void handle_status(const mavlink::mavlink_message_t *, mavlink::ardupilotmega::msg::MAG_CAL_PROGRESS &mp) {
+		auto mcs = boost::make_shared<std_msgs::UInt8>();
+
+		// How many compasses are we calibrating?
+		std::bitset<8> compass_calibrating = mp.cal_mask;
+
+		if (_rg_compass_cal_progress.size() < 3) {
+			_rg_compass_cal_progress.resize(3);
+		}
+
+		if (calibration_show.size() < 3) {
+			calibration_show.resize(3);
+			std::fill(calibration_show.begin(), calibration_show.end(), true);
+		}
```

Comment:
What would reset calibration_show?

Reply:
A new calibration request. Would you do it different?

Reply:
But where request comes from?
What would clear show flags? Because now i only see that only first time it comes it'll do initialization.

I also suggest to use array<> and define max size const.